obvs/patchscope.py

```diff
@@ -186,18 +191,34 @@ def source_forward_pass(self) -> None:
                 # TODO: validate this with non GPT2 & GPTJ models
                 self.source_model.transformer.wte.output = self.source.soft_prompt
 
-            self._source_hidden_state = self.manipulate_source().save()
-            self.source_output = self.source_model.lm_head.output[0].save()
+            self._source_hidden_state = self.manipulate_source().detach().save()
+            self.source_output = self.source_model.lm_head.output[0].detach().save()
 
     def manipulate_source(self) -> torch.Tensor:
         """
         Get the hidden state from the source representation.
 
         NB: This is seperated out from the source_forward_pass method to allow for batching.
         """
-        return getattr(getattr(self.source_model, self.MODEL_SOURCE), self.LAYER_SOURCE)[
+
+        # get the specified layer
+        layer = getattr(getattr(self.source_model, self.MODEL_SOURCE), self.LAYER_SOURCE)[
             self.source.layer
-        ].output[0][:, self._source_position, :]
+        ]
+
+        # if a head index is given, need to access the ATTN and HEAD components
+        if self.source.head is not None:
+            attn = getattr(layer, self.ATTN_SOURCE)
+            # TODO may not be .input for other models
```

Comment:
`.input` and `.output` are defined in `nnsight`. It'll be the same for all models that `nnsight` supports.

https://github.com/ndif-team/nnsight/blob/main/src/nnsight/envoy.py#L400
(note: an envoy is a wrapper around a pytorch's module)

Reply:
Depending on the implementation of the Attention, we may need to access the input or the output of the HEAD_SOURCE attribute. This is what I mean by this comment.
See also comment below.